### PR TITLE
DCP0007: Revert Treasury Expenditure Policy.

### PR DESCRIPTION
**Pull request by matheusd**
Outstanding tasks:

- [x] Define start and end times for the agenda
- [x] Link to reference implementation

### CODE REVIEW

File: dcp-0007/dcp-0007.mediawiki [hunk 1]
```diff
@@ -0,0 +1,193 @@
+<pre>
+DCP: 0007
+Title: Revert Treasury Expenditure Policy
+Author: Matheus Degiovani <contact@matheusd.com>
+Status: Active
```
Comment (davecgh) [inline dcp-0007/dcp-0007.mediawiki]:
```suggestion
Status: Defined
```

File: README.mediawiki [hunk 2]
```diff
@@ -41,4 +41,6 @@ should be performed.
 |[[./dcp-0005/dcp-0005.mediawiki|0005]]||Block Header Commitments||Active
 |-
 |[[./dcp-0006/dcp-0006.mediawiki|0006]]||Decentralized Treasury||Active
+|-
+|[[./dcp-0007/dcp-0007.mediawiki|0007]]||Revert Treasury Maximum Expenditure Policy||Proposed
```
Comment (davecgh) [inline README.mediawiki]:
```suggestion
|[[./dcp-0007/dcp-0007.mediawiki|0007]]||Revert Treasury Maximum Expenditure Policy||Defined
```

per [DCP-0000](https://github.com/decred/dcps/blob/master/dcp-0000/dcp-0000.mediawiki#specification-1).

File: dcp-0007/dcp-0007.mediawiki [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+==Reference Implementation==
+
+<pre>
+policyWindow := b.chainParams.TreasuryVoteInterval *
+    b.chainParams.TreasuryVoteIntervalMultiplier *
+    b.chainParams.TreasuryExpenditureWindow
+
+// First: sum up tspends, tadds and tbases inside the most recent
+// policyWindow.
+spentRecent, addedRecent, _, err := b.sumPastTreasuryExpenditure(preTVINode, policyWindow)
```
Comment (matheusd) [inline dcp-0007/dcp-0007.mediawiki:177]: Leaving this as an inline comment since it's not part of the spec itself, but rather discussion about its interpretation.

One thing that isn't obvious here is that TSpends and TAdds are supposed to be subject to maturity, but this is not being taken into account here. So it could happen that a tspend is being mined spending from a TAdd that "hasn't matured" yet. Note that in this case, the TAdd (which was necessarily mined in a parent block) will mature before the TSpend, so I believe "accounting-wise" this is correct, even if a little weird.

Another corollary of using this specific rule is that sending funds to the treasury immediately allows "overriding" the limit.

File: dcp-0007/dcp-0007.mediawiki [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+
+<blockquote>The sum of treasury spend amounts inside the most recent expenditure window
+cannot exceed the total amount received by the treasury in the same window, in
+addition to a 50% increase.</blockquote>
+
+==Rationale==
+
+Changing the expenditure policy to the one originally specified and approved by
+stakeholders in Politeia should not be controversial from a community discussion
+standpoint and solves the immediate issues facing the blocking of the funds by
```
Comment (davecgh) [inline dcp-0007/dcp-0007.mediawiki]:
```suggestion
standpoint and solves the immediate issues facing the blocking of spending funds by
```

File: dcp-0007/dcp-0007.mediawiki [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+the treasury.
```
Comment (davecgh) [inline dcp-0007/dcp-0007.mediawiki]:
```suggestion
the decentralized treasury.
```

File: dcp-0007/dcp-0007.mediawiki [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+
+The two policies are largely compatible at this point time, given the current
+average expenditure amount being withdrawn from the treasury, as well as the
+currently mined test transaction in mainnet being acceptable by both policies.
```
Comment (davecgh) [inline dcp-0007/dcp-0007.mediawiki]: This is really just a matter of style (plus a couple of typos), but I find this a little hard to parse.  How about the following?

```suggestion
The two policies are largely compatible at this point in time since the following relevant factors are valid under both policies:
- The current average expenditure amount being withdrawn from the treasury
- The existing <code>mainnet</code> test treasury spend transaction
```

File: dcp-0007/dcp-0007.mediawiki [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+
+{|
+!Name!!Setting
+|-
+|Deployment Version||9
+|-
+|Agenda ID||treasuryexppolicy
+|-
+|Agenda Description||Modify the treasury maximum expenditure policy as defined
+in DCP0006
```
Comment (davecgh) [inline dcp-0007/dcp-0007.mediawiki]:
```suggestion
in DCP0007
```

File: dcp-0007/dcp-0007.mediawiki [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+affected by these changes.
+
+==Reference Implementation==
+
+<pre>
+policyWindow := b.chainParams.TreasuryVoteInterval *
+    b.chainParams.TreasuryVoteIntervalMultiplier *
+    b.chainParams.TreasuryExpenditureWindow
+
+// First: sum up tspends, tadds and tbases inside the most recent
```
Comment (davecgh) [inline dcp-0007/dcp-0007.mediawiki]: I would suggest calling these out more explicitly.  I personally find the "tfoo" convention ambiguous for people not already familiar and certainly nothing else up to this point refers to `treasurybases` a `tbases`.

e.g. treasury spends, treasury adds, and treasurybases.  The difference in the spacing is so that treasurybase lines up with coinbase.

File: dcp-0007/dcp-0007.mediawiki [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+[[https://blog.decred.org/2021/06/25/Treasury-Expenditure-Policy-Bug/|Treasury Expenditure Policy Bug]]
+blog post.
+
+==Specification==
+
+Change the maximum expenditure policy to the following:
+
+<blockquote>The sum of treasury spend amounts inside the most recent expenditure window
+cannot exceed the total amount received by the treasury in the same window, in
+addition to a 50% increase.</blockquote>
```
Comment (davecgh) [inline dcp-0007/dcp-0007.mediawiki]: This needs to be more explicit, in my opinion.  Specifically, "a 50% increase" is an ambiguous amount.  e.g. Does the calculation use T-division, F-division, R-division, or C-division?  What are the exact floating point semantics (which largely differ between languages)?

All of that ambiguity can be avoided by explicitly providing the mathematical formula including the floor function to ensure integer arithmetic and positive values where the various division methods produce the same quotient.  Further, different implementations will then not be required to know the specific intricacies of the Go runtime.

File: dcp-0007/dcp-0007.mediawiki [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+addition to a 50% increase. If no amount is spent in the preceding windows,
+the maximum allowed amount equals a bootstrap amount in addition to a 50%
+increase.</blockquote>
+
+For mainnet, the expenditure policy window has a size of 6912 blocks and 6
+preceding windows are checked, for a total of 41472 blocks. A bootstrap amount
+of 16K DCR is used.
+
+A deeper overview of the issue can be read on the 
+[[https://blog.decred.org/2021/06/25/Treasury-Expenditure-Policy-Bug/|Treasury Expenditure Policy Bug]]
```
Comment (davecgh) [inline dcp-0007/dcp-0007.mediawiki]: This should probably be a reference like the other cases.

**Review by davecgh — CHANGES_REQUESTED** [review]
Some minor things I spotted.

File: dcp-0007/dcp-0007.mediawiki [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+License: CC0-1.0
+License-Code: ISC
+Requires: DCP0006
+</pre>
+
+==Abstract==
+
+This proposes reverting the currently active maximum expenditure policy of the
+treasury account to the one specified in the original Politeia proposal for the
+decentralized treasury.
```
Comment (buck54321) [inline dcp-0007/dcp-0007.mediawiki:16]: If you're "reverting" to the policy approved by stakeholders, isn't it just a bug fix? 

If a developer makes a mistake, which is arguably what has happened here (even if it was a collective decision of many developers), does there need to be a vote to fix it?

Comment (davecgh) [reply dcp-0007/dcp-0007.mediawiki:16]: > Does there need to be a vote to fix it?

Yes.  Without a vote, nodes would be hard forked from the network and havoc would ensue.  Distributed consensus requires very special care.

File: dcp-0007/dcp-0007.mediawiki (hunk 9 above)
Comment (matheusd) [reply dcp-0007/dcp-0007.mediawiki]: Github doesn't accept math in comments, so I've rendered and attached a complete formulation for the policy. Hopefully the definitions at the start are sufficient to define it unambiguously. I'll generate individual images for the terms/formulas once this is nailed down.

dcp0007-math

Comment (davecgh) [reply dcp-0007/dcp-0007.mediawiki]: That looks much better.  The only part that is still slightly ambiguous to me is what happens when `h < pw`?  Is it clamped to 0, or considered illegal resulting in `S_{spent} = 0` and `S_{added} = 0`?  I realize this doesn't matter on mainnet, but it will on simnet.

I believe the correct behavior should be clamping it zero, in which case you probably want:

`S_{spent} = \sum_{i=max(h-pw, 0)}^{h-1} \beta_{i}^{ts}`

image

Similarly for `S_{added}`.

Also, I noticed `S_{spent}` in the equation and `S_{spend}` in the informal description.

Comment (matheusd) [reply dcp-0007/dcp-0007.mediawiki]: Updated.

image

Comment (JoeGruffins) [reply dcp-0007/dcp-0007.mediawiki]: With my crude math, on mainnet, it looks like this allows for spending per window, not counting treasury adds:

12,948dcr now
10,933dcr next year
5,502dcr in five years
2,361dcr in ten years
430dcr in twenty years

Look right?

File: dcp-0007/dcp-0007.mediawiki [hunk 12]
```diff
@@ -0,0 +1,194 @@
...
+
+==Motivation==
+
+The current algorithm that calculates the maximum expenditure allowed by a set
+of treasury spend transactions within a given window of blocks takes into
+account the average spent in a preceding window.
+
+This is different then the policy that was originally proposed and approved by
```
Comment (JoeGruffins) [inline dcp-0007/dcp-0007.mediawiki]: different than

File: dcp-0007/dcp-0007.mediawiki (hunk 9 above)
Comment (matheusd) [reply dcp-0007/dcp-0007.mediawiki]: Those are about right, yeah.

Comment (matheusd) [reply dcp-0007/dcp-0007.mediawiki]: Updated to use svg renderings of the equations.

**Review by JoeGruffins — APPROVED** [review]

**Review by dnldd — APPROVED** [review]
Reads well, should be good once the placeholders for the reference implementations are updated.

**Comment by xaur** [comment]
In layman terms, how will this new algorithm handle exchange rate drop to $10 or even $1?

**Comment by davecgh** [comment]
> In layman terms, how will this new algorithm handle exchange rate drop to $10 or even $1?

It doesn't and neither does the current buggy version either.  The reality is that no policy is going to be perfect and, as the blog post on this topic called out, future changes are absolutely going to be necessary after more data on actual usage has been gathered.

Quoting it:

> As demonstrated, imposing algorithmic spending constraints that are flexible enough to handle ongoing operations as well as sporadic high variance payments while simultaneously avoiding malicious actors is a delicate balance to get right and any proposed solution might have subtle failure modes that will need to be addressed.
>
> Future work will need to refine spending constraints using information about spending patterns that emerge through continued use of the decentralized treasury.

Also, it's not a new algorithm, rather, it is the one originally proposed in the Pi proposal, which was also called out as something that would need to be changed.

File: dcp-0007/dcp-0007.mediawiki (hunk 9 above)
Comment (rstaudt2) [reply dcp-0007/dcp-0007.mediawiki]: One thing that is somewhat obvious, but might be worth calling out, is whether the "sum of the transactions" refers to the sum of the inputs or outputs in each case (e.g., in the case of treasury add and treasury base, the sum refers to the transaction outputs being added to the treasury, but in the case of treasury spends the sum refers to the transaction inputs being spent from the treasury which includes fees, [as accounted for in your draft reference implementation](https://github.com/decred/dcrd/blob/1b04ee6c1f4569d53ba398ea2e6e20478dadc71f/blockchain/treasury.go#L631)).

Comment (matheusd) [reply dcp-0007/dcp-0007.mediawiki]: Tweaked the relevant section to the below. Do you think this is sufficient?

tx_types

Comment (rstaudt2) [reply dcp-0007/dcp-0007.mediawiki]: Yeah, that looks good to me.

**Comment by xaur** [comment]
> It doesn't and neither does the current buggy version either.

I think I see it now and can answer my own question "how will this change react to a sharp drop of DCR exchange rate?" with some napkin math.

For simplicity, let's assume (a) DCR is $100, (b) new treasury gains 10K DCR per month, (c) spends USD 300K per month (3,000 DCR), (d) conditions a-c hold for several months. Current policy based on past _spent_ amount yields monthly allowance of around 4,500 DCR. Proposed policy based on past _added_ amount translates to monthly allowance of around 15,000 DCR.

Now, if the rate collapses to $10 we will need to burn 30K DCR to keep things running and both policies will not allow that. But the proposed one will perform much better with significantly higher limit (allowing to cut _less_ things). Also, since it includes TADD tx, it allows extra flexibility by sending some DCR from the legacy treasury and further increasing the allowance.

It looks way smarter than I thought :)

**Review by davecgh — COMMENTED** [review]
It looks like this will be good to go once the agenda times and reference link are updated as well as commits squashed.

File: dcp-0007/dcp-0007.mediawiki [hunk 13]
```diff
@@ -0,0 +1,239 @@
...
+
+This agenda will be deployed to mainnet using the standard Decred on-chain
+voting infrastructure as follows:
+
+{|
+!Name!!Setting
+|-
+|Deployment Version||9
+|-
+|Agenda ID||treasuryexppolicy
```
Comment (davecgh) [inline dcp-0007/dcp-0007.mediawiki]: This should be `reverttreasurypolicy`.

**Review by davecgh — CHANGES_REQUESTED** [review]

File: dcp-0007/dcp-0007.mediawiki [hunk 14]
```diff
@@ -0,0 +1,239 @@
...
+|-
+|Agenda Description||Modify the treasury maximum expenditure policy as defined
```
Comment (davecgh) [inline dcp-0007/dcp-0007.mediawiki]:
```suggestion
|Agenda Description||Change the treasury maximum expenditure policy as defined
```

It's `Change` in the code instead of `Modify`.

**Review by davecgh — APPROVED** [review]
